services/storehost/extStats.go

```diff
@@ -0,0 +1,434 @@
+					deleteExtents = append(deleteExtents, id)
+				}
+
+				time.Sleep(pause) // take a short nap
+
+				t.RLock()
+			}
+
+			t.RUnlock()
+
+			// get lock exclusive, and delete extents from the map whose ref has dropped to zero
+			t.Lock()
+			for _, id := range deleteExtents {
+				if ctx, ok := t.extents[string(id)]; ok && atomic.LoadInt64(&ctx.ref) == 0 {
+					delete(t.extents, id)
+				}
+			}
+			t.Unlock()
+
+		case <-t.stopC:
+			t.logger.Info("extStatsReporter: schedulerPump stopped")
+			break pump
+		}
+
+		if newInterval := reportInterval.Load().(time.Duration); interval != newInterval {
+			ticker.Stop()
+			interval = newInterval
+			ticker = time.NewTicker(interval)
+
+			t.logger.WithFields(bark.Fields{
+				`report-interval`: interval,
+			}).Info("extStatsReporter: schedulerPump: report-interval changed")
+		}
+	}
+}
+
+func (t *ExtStatsReporter) extentOpened(id uuid.UUID, ext *extentContext, intent OpenIntent) {
+
+	if intent != OpenIntentAppendStream &&
+		intent != OpenIntentSealExtent &&
+		intent != OpenIntentPurgeMessages &&
+		intent != OpenIntentReplicateExtent {
+
+		// t.logger.WithFields(bark.Fields{ // #perfdisable
+		// 	common.TagExt: id,     // #perfdisable
+		// 	`intent`:      intent, // #perfdisable
+		// }).Info("extStatsReporter: ignoring extent opened event") // #perfdisable
+
+		return // ignore, if not any of the interesting 'intents'
+	}
+
+	// get exclusive lock, since we could be adding to the map
+	t.Lock()
+	defer t.Unlock()
+
+	ctx, ok := t.extents[string(id)]
+
+	if ok && ctx.ext == ext {
+
+		atomic.AddInt64(&ctx.ref, 1) // add ref
+
+	} else {
+
+		// assert(ctx.ref == 0) //
+
+		// create a new context
+		ctx = &extStatsContext{ext: ext, ref: 1}
+		t.extents[string(id)] = ctx
+	}
+
+	// t.logger.WithFields(bark.Fields{ // #perfdisable
+	// 	common.TagExt: id,                         // #perfdisable
+	// 	`intent`:      intent,                     // #perfdisable
+	// 	`ctx.ref`:     atomic.LoadInt64(&ctx.ref), // #perfdisable
+	// }).Info("extStatsReporter: extent opened") // #perfdisable
+
+	return
+}
+
+func (t *ExtStatsReporter) extentClosed(id uuid.UUID, ext *extentContext, intent OpenIntent) (done bool) {
+
+	if intent != OpenIntentAppendStream &&
+		intent != OpenIntentSealExtent &&
+		intent != OpenIntentPurgeMessages &&
+		intent != OpenIntentReplicateExtent {
+
+		// t.logger.WithFields(bark.Fields{ // #perfdisable
+		// 	common.TagExt: id,     // #perfdisable
+		// 	`intent`:      intent, // #perfdisable
+		// }).Info("extStatsReporter: ignoring extent closed event") // #perfdisable
+
+		return true // ignore, if not any of the interesting 'intents'
+	}
+
+	// get lock shared, when reading through the map
+	t.RLock()
+	defer t.RUnlock()
+
+	ctx, ok := t.extents[string(id)]
+
+	if ok && ctx.ext == ext {
+
+		// remove ref, if it drops to zero, then send out one last report
+		if atomic.AddInt64(&ctx.ref, -1) == 0 {
+			t.trySendReport(id, ext, ctx)
```

Comment:
we filter in calls that only 'mutate' the extent .. so we will not send out reports in response to GAFT calls.